### PR TITLE
Integrate Limited Relays

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For https://github.com/libp2p/go-libp2p/issues/1039

### CODE REVIEW

**Comment by aarshkshah1992** [comment]
@vyzo 

1. It's a breaking change (we don't want to support v1s anymore)
2. Should we use the Reservation  returned by the v2 client for anything at all in AutoRelay  ?

**Comment by vyzo** [comment]
1. v2 is backwards compatible for dialing/listening to v1, so no it's not a breaking change (other than the relay options disappearing)
2. The reservation provides two things you want to use (voucher aside):
    - the public relay addrs which you want to advertise (they are complete, with p2p-circuit in them so you just advertise them as they are)
    - the expiration of the reservation so that you know when to refresh; this is in unix time (seconds).

File: config/config.go [hunk 1]
```diff
@@ -25,8 +25,7 @@ import (
 
 	autonat "github.com/libp2p/go-libp2p-autonat"
 	blankhost "github.com/libp2p/go-libp2p-blankhost"
-	circuit "github.com/libp2p/go-libp2p-circuit"
-	discovery "github.com/libp2p/go-libp2p-discovery"
+	circuitv2client "github.com/libp2p/go-libp2p-circuit/v2/client"
```
Comment (vyzo) [inline config/config.go:28]: that's a huge name; just `circuitv2` will do.

File: config/config.go [hunk 2]
```diff
@@ -92,7 +90,7 @@ type Config struct {
 
 	EnableAutoRelay bool
 	AutoNATConfig
-	StaticRelays []peer.AddrInfo
+	StaticV2Relays []peer.AddrInfo
```
Comment (vyzo) [inline config/config.go:93]: no need to rename this variable.

File: config/config.go (hunk 1 above)
Comment (vyzo) [reply config/config.go:28]: Even just `circuit` will do in fact.

File: options.go [hunk 3]
```diff
@@ -230,41 +226,34 @@ func DisableRelay() Option {
...
 
-// EnableAutoRelay configures libp2p to enable the AutoRelay subsystem.
+// EnableAutoRelay configures libp2p to enable the AutoRelay subsystem using statically configured v2 Relays.
 //
 // Dependencies:
 //  * Relay (enabled by default)
-//  * Routing (to find relays), or StaticRelays/DefaultStaticRelays.
-//
-// This subsystem performs two functions:
+//  * StaticV2Relays/DefaultStaticV2Relays.
```
Comment (vyzo) [inline options.go:233]: drop the v2 from these names.

File: p2p/host/relay/autorelay.go [hunk 4]
```diff
@@ -30,15 +31,9 @@ var (
...
-	"/ip4/147.75.80.110/tcp/4001/p2p/QmbFgm5zan8P6eWWmeyfncR5feYEMPbht5b1FW1C37aQ7y",
-	"/ip4/147.75.80.110/udp/4001/quic/p2p/QmbFgm5zan8P6eWWmeyfncR5feYEMPbht5b1FW1C37aQ7y",
-	"/ip4/147.75.195.153/tcp/4001/p2p/QmW9m57aiBDHAkKj9nmFSEn7ZqrcF1fZS4bipsTCHburei",
-	"/ip4/147.75.195.153/udp/4001/quic/p2p/QmW9m57aiBDHAkKj9nmFSEn7ZqrcF1fZS4bipsTCHburei",
-	"/ip4/147.75.70.221/tcp/4001/p2p/Qme8g49gm3q4Acp7xWBKg3nAa9fxZ1YmyDJdyGgoG6LsXh",
-	"/ip4/147.75.70.221/udp/4001/quic/p2p/Qme8g49gm3q4Acp7xWBKg3nAa9fxZ1YmyDJdyGgoG6LsXh",
-}
+// These are the known PL-operated V2 relays
+// TODO Fill this up
+var DefaultV2Relays = []string{}
```
Comment (vyzo) [inline p2p/host/relay/autorelay.go:36]: just `DefaultRelays`.

File: p2p/host/relay/autorelay.go [hunk 5]
```diff
@@ -196,16 +191,13 @@ func (ar *AutoRelay) tryRelay(ctx context.Context, pi peer.AddrInfo) bool {
 		return false
 	}
 
-	ok, err := circuit.CanHop(ctx, ar.host, pi.ID)
+	rsv, err := circuitv2client.Reserve(ctx, ar.host, pi)
```
Comment (vyzo) [inline p2p/host/relay/autorelay.go:194]: just `circuit` no need for comically long names. Also, `rsv` should be `rsvp`.

File: p2p/host/relay/autorelay.go [hunk 6]
```diff
@@ -196,16 +191,13 @@ func (ar *AutoRelay) tryRelay(ctx context.Context, pi peer.AddrInfo) bool {
...
-		log.Debugf("error querying relay: %s", err.Error())
+		log.Debugf("error reserving slot on relay: %s", err.Error())
 		return false
 	}
 
-	if !ok {
-		// not a hop relay
-		return false
-	}
+	ar.host.Peerstore().AddAddrs(pi.ID, rsv.Relay.Addrs, peerstore.ConnectedAddrTTL)
```
Comment (vyzo) [inline p2p/host/relay/autorelay.go:200]: no, this has changed in the v2 code.
The addresses that you get are your own relay addresses, and they are in the `rsv(p).Addrs` field.

Comment (vyzo) [reply p2p/host/relay/autorelay.go:200]: You also need to schedule reservation refresh, use the `Expiration` field in the reservation to make the determination.

File: config/config.go (hunk 1 above)
Comment (aarshkshah1992) [reply config/config.go]: done.

File: options.go (hunk 3 above)
Comment (aarshkshah1992) [reply options.go:233]: done.

File: config/config.go (hunk 2 above)
Comment (aarshkshah1992) [reply config/config.go]: done.

File: p2p/host/relay/autorelay.go (hunk 4 above)
Comment (aarshkshah1992) [reply p2p/host/relay/autorelay.go]: done.

File: p2p/host/relay/autorelay.go (hunk 5 above)
Comment (aarshkshah1992) [reply p2p/host/relay/autorelay.go]: done.

File: p2p/host/relay/autorelay.go (hunk 6 above)
Comment (aarshkshah1992) [reply p2p/host/relay/autorelay.go]: done.

**Comment by aarshkshah1992** [comment]
@vyzo 

1. Have addressed your feedback.
2. We refresh reservations now if the Expiry TTL is above a reasonable threshold.

File: p2p/host/relay/autorelay.go [hunk 7]
```diff
@@ -127,6 +139,39 @@ func (ar *AutoRelay) background(ctx context.Context) {
...
+				continue
+			}
+
+			ar.mx.Lock()
+			info := &relayInfo{
+				addrs: rsvp.Addrs,
+			}
+			if rsvp.LimitDuration > MinTTLRefresh {
+				info.resfreshAt = time.Now().Add(rsvp.LimitDuration / 2)
+			}
```
Comment (vyzo) [inline p2p/host/relay/autorelay.go:166]: this is completely wrong -- the limit duration is the actual relay limit; you want to look at the `Expiration` which is when your reservation expires.

File: p2p/host/relay/autorelay.go [hunk 8]
```diff
@@ -214,7 +254,17 @@ func (ar *AutoRelay) tryRelay(ctx context.Context, pi peer.AddrInfo) bool {
...
-	ar.relays[pi.ID] = struct{}{}
+
+	rinfo := &relayInfo{
+		addrs: rsvp.Addrs,
+	}
+	// If it's a limited time Relay, we need to schedule a reservation refresh before the reservation expires.
+	// However, avoid scheduling a refresh if the TTL is less than 10 minutes to prevent repeated refreshes.
+	if rsvp.LimitDuration > MinTTLRefresh {
+		rinfo.resfreshAt = time.Now().Add(rsvp.LimitDuration / 2)
+	}
```
Comment (vyzo) [inline p2p/host/relay/autorelay.go:265]: same thing here.

**Review by vyzo — CHANGES_REQUESTED** [review]

**Comment by vyzo** [comment]
closing as we'll have to do this in pieces, and probably have a transition period for autorelay where we support both.

**Pull request closed without merging**